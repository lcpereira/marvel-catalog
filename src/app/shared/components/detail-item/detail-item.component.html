<div class="container">
  <div class="detaild-header d-flex justify-content-between mb-2">
    <span class="mdi mdi-arrow-left back" (click)="backPage()"></span>
    <span class="mdi mdi-heart favorite"></span>
  </div>
  <div *ngIf="item$ | async as item; else loadingTemplate">
    <div class="detail-item row">
      <div class="col-md-4 col-sm-6 mb-4">
        <img class="thumb" [src]="item.thumbnail.path + '.' + item.thumbnail.extension" />
      </div>
      <div class="col-md-8 col-sm-6">
        <h2>
          <strong class="title">{{ item.title || item.name }}</strong>
        </h2>
        <p class="description mt-4">{{ item.description || 'Sem descrição' }}</p>
        <div class="mt-2" *ngIf="item.creators?.items?.length">
          <h4><strong>Creators</strong></h4>
          <table class="table">
            <thead>
              <tr>
                <th scope="col">Name</th>
                <th scope="col">Role</th>
              </tr>
            </thead>
            <tbody>
              <tr *ngFor="let creator of item.creators.items">
                <td>{{ creator.name }}</td>
                <td>{{ creator.role }}</td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>

<ng-template #loadingTemplate> Carregando... </ng-template>
