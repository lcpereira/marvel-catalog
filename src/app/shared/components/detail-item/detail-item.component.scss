.detaild-header {
  font-size: 30px;
  color: #aaaaaa;
  & .back, & .favorite {
    cursor: pointer;
    &:hover {
      color: #3a3a3a;
    }
    &.active {
      color: #f56a6a !important;
    }
  }

}

.detail-item {
  & .thumb {
    width: 100%;
    height: auto;
    max-width: 800px;
    border-radius: 8px;
  }

  & .detail-item-content {
    padding: 10px;

    & .title, & .description {
      display: -webkit-box;
      -webkit-box-orient: vertical;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    & .title {
      -webkit-line-clamp: 1;
    }

    & .description {
      font-size: 14px;
      -webkit-line-clamp: 2;
    }
  }
}
