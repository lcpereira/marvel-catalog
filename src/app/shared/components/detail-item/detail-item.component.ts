import { MarvelService } from './../../services/marvel.service';
import { MarvelType } from './../../enums/marvel-type.enum';
import { Component, OnDestroy, OnInit } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { Subscription, Observable } from 'rxjs';

@Component({
  selector: 'ngx-detail-item',
  templateUrl: 'detail-item.component.html',
  styleUrls: ['detail-item.component.scss'],
})
export class DetailItemComponent implements OnInit, OnDestroy {
  type: MarvelType;
  itemId: string;
  item$: Observable<any>;

  private subs = new Subscription();

  constructor(
    private router: Router,
    private route: ActivatedRoute,
    private marvelService: MarvelService
  ) {}

  ngOnInit(): void {
    this.subs.add(this.route.data.subscribe((data) => (this.type = data.type)));

    this.subs.add(
      this.route.params.subscribe((params) => {
        this.itemId = params.id;
        this.loadItem();
      })
    );
  }

  loadItem(): void {
    if (this.type === MarvelType.COMICS) {
      this.item$ = this.marvelService.getComic(this.itemId);
    } else {
      this.item$ = this.marvelService.getCharacter(this.itemId);
    }
  }

  backPage(): void {
    if (this.type === MarvelType.COMICS) {
      this.router.navigate(['/comics']);
    } else {
      this.router.navigate(['/characters']);
    }
  }

  ngOnDestroy(): void {
    this.subs.unsubscribe();
  }
}
