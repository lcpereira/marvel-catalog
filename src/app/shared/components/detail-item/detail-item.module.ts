import { RouterModule } from '@angular/router';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DetailItemComponent } from './detail-item.component';

@NgModule({
  declarations: [DetailItemComponent],
  exports: [DetailItemComponent],
  imports: [CommonModule, RouterModule],
})
export class DetailItemModule {}
