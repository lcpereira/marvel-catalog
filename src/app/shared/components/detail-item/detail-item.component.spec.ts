import { MarvelType } from './../../enums/marvel-type.enum';
import { RouterTestingModule } from '@angular/router/testing';
import { MarvelService } from './../../services/marvel.service';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { waitForAsync, ComponentFixture, TestBed } from '@angular/core/testing';

import { DetailItemComponent } from './detail-item.component';
import { Router } from '@angular/router';
import { Component } from '@angular/core';

@Component({
  selector: 'ngx-fake-component',
  template: ``,
})
export class FakeComponent {}

describe('DetailItemComponent', () => {
  let component: DetailItemComponent;
  let router: Router;
  let fixture: ComponentFixture<DetailItemComponent>;

  beforeEach(
    waitForAsync(() => {
      TestBed.configureTestingModule({
        declarations: [DetailItemComponent, FakeComponent],
        imports: [
          HttpClientTestingModule,
          RouterTestingModule.withRoutes([
            { path: 'comics', component: FakeComponent },
            { path: 'characters', component: FakeComponent },
          ]),
        ],
        providers: [MarvelService],
      }).compileComponents();
    })
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(DetailItemComponent);
    component = fixture.componentInstance;
    router = TestBed.inject(Router);
    fixture.detectChanges();
  });

  it('should create the app', () => {
    expect(component).toBeTruthy();
  });

  it('should load comic item', () => {
    component.itemId = '123';
    component.type = MarvelType.COMICS;

    component.loadItem();

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should load character item', () => {
    component.itemId = '456';
    component.type = MarvelType.CHARACTERS;

    component.loadItem();

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should back to comics page', () => {
    jest.spyOn(router, 'navigate').mockReturnValue(null);

    component.type = MarvelType.COMICS;

    component.backPage();

    expect(router.navigate).toHaveBeenCalledWith(['/comics']);
  });

  it('should back to characters page', () => {
    jest.spyOn(router, 'navigate').mockReturnValue(null);

    component.type = MarvelType.CHARACTERS;

    component.backPage();

    expect(router.navigate).toHaveBeenCalledWith(['/characters']);
  });
});
