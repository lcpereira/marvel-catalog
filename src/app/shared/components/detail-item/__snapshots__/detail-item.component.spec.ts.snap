// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailItemComponent should load character item 1`] = `
<ngx-detail-item
  item$={[Function Observable]}
  itemId={[Function String]}
  marvelService={[Function MarvelService]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  subs={[Function Subscription]}
  type={[Function String]}
>
  <div
    class="container"
  >
    <div
      class="detaild-header d-flex justify-content-between mb-2"
    >
      <span
        class="mdi mdi-arrow-left back"
      />
      <span
        class="mdi mdi-heart favorite"
      />
    </div>
    
     Carregando... 
  </div>
</ngx-detail-item>
`;

exports[`DetailItemComponent should load comic item 1`] = `
<ngx-detail-item
  item$={[Function Observable]}
  itemId={[Function String]}
  marvelService={[Function MarvelService]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  subs={[Function Subscription]}
  type={[Function String]}
>
  <div
    class="container"
  >
    <div
      class="detaild-header d-flex justify-content-between mb-2"
    >
      <span
        class="mdi mdi-arrow-left back"
      />
      <span
        class="mdi mdi-heart favorite"
      />
    </div>
    
     Carregando... 
  </div>
</ngx-detail-item>
`;
